src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/ModelMetadataIdentity.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Reflection;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Mvc.ModelBinding.Metadata
+{
+    /// <summary>
+    /// A key type which identifies a <see cref="ModelMetadata"/>.
+    /// </summary>
+    public struct ModelMetadataIdentity
+    {
+        /// <summary>
+        /// Creates a <see cref="ModelMetadataIdentity"/> for the provided model <see cref="Type"/>.
+        /// </summary>
+        /// <param name="modelType">The model <see cref="Type"/>.</param>
+        /// <returns>A <see cref="ModelMetadataIdentity"/>.</returns>
+        public static ModelMetadataIdentity ForType([NotNull] Type modelType)
+        {
+            return new ModelMetadataIdentity()
+            {
+                ModelType = modelType,
+            };
+        }
+
+        /// <summary>
+        /// Creates a <see cref="ModelMetadataIdentity"/> for the provided <see cref="ParameterInfo"/>.
+        /// </summary>
+        /// <param name="parameterInfo">The model parameter.</param>
+        /// <returns>A <see cref="ModelMetadataIdentity"/>.</returns>
+        public static ModelMetadataIdentity ForParameter([NotNull] ParameterInfo parameterInfo)
+        {
+            return new ModelMetadataIdentity()
+            {
+                ParameterInfo = parameterInfo,
+
+                Name = parameterInfo.Name,
+                ModelType = parameterInfo.ParameterType,
+            };
+        }
+
+
+        /// <summary>
+        /// Creates a <see cref="ModelMetadataIdentity"/> for the provided property.
+        /// </summary>
+        /// <param name="modelType">The model type.</param>
+        /// <param name="name">The name of the property.</param>
+        /// <param name="containerType">The container type of the model property.</param>
+        /// <returns>A <see cref="ModelMetadataIdentity"/>.</returns>
+        public static ModelMetadataIdentity ForProperty(
+            [NotNull] Type modelType,
+            [NotNull] string name,
+            [NotNull] Type containerType)
+        {
+            if (string.IsNullOrEmpty(name))
```

Comment:
nit: we tend to leave out `[NotNull]` when doing a `IsNullOrEmpty` check.  that ensures `null` and empty cases result in the same `Message`